File: protelis-interpreter/src/main/java/org/protelis/lang/datatype/JVMEntity.java
/*
 * Copyright (C) 2021, Danilo Pianini and contributors listed in the project's build.gradle.kts or pom.xml file.
 *
 * This file is part of Protelis, and is distributed under the terms of the GNU General Public License,
 * with a linking exception, as described in the file LICENSE.txt in this project's top directory.
 */

package org.protelis.lang.datatype;

import java.io.Serializable;
import java.lang.reflect.Method;
import java.lang.reflect.Modifier;

import org.eclipse.xtext.common.types.JvmFeature;
import org.eclipse.xtext.common.types.JvmField;
import org.eclipse.xtext.common.types.JvmOperation;

import edu.umd.cs.findbugs.annotations.SuppressFBWarnings;

/**
 * A runtime Protelis view of a {@link JvmFeature}.
 *
 */
public final class JVMEntity implements Serializable {

    private static final long serialVersionUID = 1;
    private final String memberName;
    private final SupportedEntityTypes memberType;
    private final String typeName;
    @SuppressFBWarnings("SE_TRANSIENT_FIELD_NOT_RESTORED")
    private transient Object value;
    @SuppressFBWarnings("SE_TRANSIENT_FIELD_NOT_RESTORED")
    private transient Class<?> type;

    /**
     * @param feature the feature to be used at runtime
     */
    public JVMEntity(final JvmFeature feature) {
        typeName = feature.getDeclaringType().getQualifiedName();
        memberName = feature.getSimpleName();
        if (feature instanceof JvmField) {
            memberType = SupportedEntityTypes.FIELD;
            value = ((JvmField) feature).getConstantValue();
        } else if (feature instanceof JvmOperation) {
            memberType = SupportedEntityTypes.METHOD;
        } else {
            throw new IllegalArgumentException("Unknown JvmFeature type " + feature.getClass() + " (" + feature + ')');
        }
    }

    /**
     * @param method the {@link Method} that this {@link JVMEntity} represents
     */
    public JVMEntity(final Method method) {
        typeName = method.getDeclaringClass().getName();
        memberName = method.getName();
        memberType = SupportedEntityTypes.METHOD;
    }

    /**
     * @return the simple name of this entity
     */
    public String getMemberName() {
        return memberName;
    }

    /**
     * @return The type this entity belongs to
     */
    @SuppressFBWarnings(value = "EI_EXPOSE_REP", justification = "The field is immutable")
    public Class<?> getType() {
        if (type == null) {
            try {
                type = Class.forName(typeName);
            } catch (ClassNotFoundException e) {
                throw new IllegalStateException(e);
            }
        }
        return type;
    }

    /**
     * @return the resolution of the entity: returns this in case of callable entity
     *         (such as a method), and the value of the referenced field otherwise.
     */
    public Object getValue() {
        if (value == null) {
            try {
                switch (memberType) {
                    case FIELD:
                        final java.lang.reflect.Field field = getType().getField(memberName);
                        if (Modifier.isStatic(field.getModifiers())) {
                            value = field.get(null);
                        } else {
                            throw new IllegalStateException("Protelis cannot access non-static fields such as " + field);
                        }
                        break;
                    case METHOD:
                        value = this;
                        break;
                    default: throw new  IllegalStateException("Fix Protelis code, it's bugged.");
                }
            } catch (NoSuchFieldException | SecurityException | IllegalAccessException e) {
                throw new IllegalStateException(e);
            }
        }
        return value;
    }

    @Override
    public String toString() {
        return memberType + ":" + typeName + "." + memberName;
    }

    enum SupportedEntityTypes {
        FIELD,
        METHOD
    }

}

File: protelis-interpreter/src/main/kotlin/org/protelis/lang/datatype/JVMEntity.kt
package org.protelis.lang.datatype

import org.eclipse.xtext.common.types.JvmFeature
import org.eclipse.xtext.common.types.JvmField
import org.eclipse.xtext.common.types.JvmOperation
import java.io.Serializable
import java.lang.reflect.Method
import java.lang.reflect.Modifier

/**
 * A runtime Protelis view of a [JvmFeature].
 */
data class JVMEntity(
    private val memberName: String,
    private val memberType: SupportedEntityTypes,
    private val typeName: String,
) : Serializable {

    private companion object {
        private const val serialVersionUID = 1L
    }

    @Transient
    private var value: Any? = null

    @Transient
    private var type: Class<*>? = null

    constructor(feature: JvmFeature) : this(
        feature.simpleName,
        when (feature) {
            is JvmField -> SupportedEntityTypes.FIELD
            is JvmOperation -> SupportedEntityTypes.METHOD
            else -> error("Unknown JvmFeature type ${feature::class.java} ($feature)")
        },
        feature.declaringType.qualifiedName,
    ) {
        if (feature is JvmField) {
            value = feature.constantValue
        }
    }

    constructor(method: Method) : this(
        method.name,
        SupportedEntityTypes.METHOD,
        method.declaringClass.name,
    )

    /**
     * @return the simple name of this entity
     */
    fun getMemberName(): String = memberName

    /**
     * @return The type this entity belongs to
     */
    fun getType(): Class<*> = type ?: try {
        Class.forName(typeName).also { type = it }
    } catch (e: ClassNotFoundException) {
        error(e)
    }

    /**
     * @return the resolution of the entity: returns this in case of callable entity
     * (such as a method), and the value of the referenced field otherwise.
     */
    fun getValue(): Any? {
        if (value == null) {
            value = try {
                when (memberType) {
                    SupportedEntityTypes.FIELD -> {
                        val field = getType().getField(memberName)
                        if (Modifier.isStatic(field.modifiers)) {
                            field.get(null)
                        } else {
                            error("Protelis cannot access non-static fields such as $field")
                        }
                    }
                    SupportedEntityTypes.METHOD -> this
                }
            } catch (e: NoSuchFieldException) {
                error(e)
            } catch (e: SecurityException) {
                error(e)
            } catch (e: IllegalAccessException) {
                error(e)
            }
        }
        return value
    }

    override fun toString(): String = "$memberType:$typeName.$memberName"

    /**
     * Represents the type of a [JVMEntity].
     */
    enum class SupportedEntityTypes {
        /** Represents a field in a JVM entity. */
        FIELD,

        /** Represents a method in a JVM entity. */
        METHOD,
    }
}
